src/daemon/scheduler.rs

```diff
@@ -169,230 +126,161 @@ fn make_cas_event_triggers(event_queue: Arc<MessageQueue>, ca_manager: Arc<CaMan
+                .sync_repo(ca.handle().clone(), self.config.ca_refresh_start_up(use_jitter));
+
+            // If suspension is enabled then plan a task for it. Since this is
+            // a cheap no-op in most cases, we do not need jitter. If we do not
+            // add this task then it will not be executed (obviously), but more
+            // importantly.. by adding this task we ensure that it will keep being
+            // re-scheduled when it's done.
+            if self.config.suspend_child_after_inactive_seconds().is_some() {
+                self.tasks.suspend_children(ca.handle().clone(), now())
             }
-        })
-    })
-}
+        }
 
-#[cfg(feature = "multi-user")]
-fn make_login_cache_sweeper_sh(cache: Arc<LoginSessionCache>) -> ScheduleHandle {
-    SkippingScheduler::run(60, "sweep session decryption cache", move || {
-        let rt = Runtime::new().unwrap();
-        rt.block_on(async {
-            if let Err(e) = cache.sweep() {
-                error!("Background sweep of session decryption cache failed: {}", e);
-            }
-        })
-    })
-}
+        self.tasks.republish_if_needed(now());
+        self.tasks.refresh_announcements_info(now());
 
-struct SkippingScheduler;
-
-impl SkippingScheduler {
-    fn run<F>(seconds: u32, name: &'static str, f: F) -> ScheduleHandle
-    where
-        F: FnMut() + Clone + Send + 'static,
-    {
-        let lock = RunLock::new();
-
-        let mut scheduler = clokwerk::Scheduler::new();
-        scheduler.every(seconds.seconds()).run(move || {
-            if lock.is_running() {
-                warn!(
-                    "Previous background job '{}' is still running, will skip and try again in {} seconds",
-                    name, seconds
-                )
-            } else {
-                lock.run();
-                let mut f = f.clone();
-                f();
-                lock.done();
-            }
-        });
+        #[cfg(feature = "multi-user")]
+        self.tasks.sweep_login_cache(in_minutes(1));
 
-        scheduler.watch_thread(Duration::from_millis(100))
+        Ok(())
     }
-}
-
-struct RunLock {
-    state: RwLock<RunState>,
-}
 
-impl RunLock {
-    fn new() -> Self {
-        RunLock {
-            state: RwLock::new(RunState(false)),
-        }
-    }
+    async fn sync_repo(&self, ca: Handle) {
+        debug!("Synchronize CA {} with repository", ca);
 
-    fn run(&self) {
-        self.state.write().unwrap().run();
-    }
+        if let Err(e) = self.ca_manager.cas_repo_sync_single(&ca).await {
+            let next = self.config.requeue_remote_failed();
 
-    fn done(&self) {
-        self.state.write().unwrap().done();
-    }
+            error!(
+                "Failed to publish for '{}'. Will reschedule to: '{}'. Error: {}",
+                ca, next, e
+            );
 
-    fn is_running(&self) -> bool {
-        self.state.read().unwrap().is_running()
+            self.tasks.sync_repo(ca, next);
```

Comment:
Should this also be done in an else branch, like in `sync_parent()` below?

Reply:
no, in this case we only reschedule if something failed. If another change happens, e.g. MFT/CRL is re-issued or ROAs are changed - then that event will trigger that a new repo sync is scheduled.

For the parent exchange we want to reschedule soon in case of a failure, and later (24-36 hours by default) in case the exchange was successful. Parent syncs are not triggered by other internal events (wel except for repo configuration changes) - so they need to be planned here (and at startup).